system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "subscriber_topic_statistics.hpp"
+#include "system_metrics_collector/constants.hpp"
+#include "system_metrics_collector/metrics_message_publisher.hpp"
+
+using metrics_statistics_msgs::msg::MetricsMessage;
+using std::placeholders::_1;
+
+namespace topic_statistics_collector
+{
+template<typename T>
```

Comment:
How will this work, putting template implementations in the `.cpp` file? I don't see any explicit template instantiations in this `.cpp` file (and that's not the way to go if you want users to be able to use any type `T`), so you'll be able to compile this translation unit successfully but you're eventually going to get undefined symbols errors from the linker. All of this stuff needs to be moved to the `.hpp` file.

Reply:
@mm318 This is ok, you can put templates in CPP files as long as its' not used in another compilation unit.
What we should not do, is have the declaration (w/o the impl in a header)

Reply:
Maybe I'm misunderstanding this class, but I thought this class will be used in another compilation unit, a ROS2 developer's compilation unit.

Reply:
Switched to single file as per the [coding standard](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-contained).